From 5db0bd65b422b26f7392e910a406c71d851b5bdb Mon Sep 17 00:00:00 2001
From: Raviteja Sunkara <raviteja@codeaurora.org>
Date: Mon, 8 Aug 2011 16:03:48 +0530
Subject: [PATCH] Show all the ancestors of a project in
 Admin->Projects->Access Screen

With this commit, the "Rights Inherit From" section lists all of the
project's ancestors instead of just the immediate parent. This helps
to easily know list of projects from which a project inherits the rights.

Bug: issue 903
Change-Id: I3e1c031f45a8962ef8c90a5f89077ae51512b6d2
---
 .../google/gerrit/common/data/ProjectAccess.java   |    9 --
 .../gerrit/common/data/ProjectAdminService.java    |    9 ++-
 .../gerrit/client/admin/ProjectAccessEditor.java   |   29 +++++--
 .../gerrit/client/admin/ProjectAccessEditor.ui.xml |    2 +-
 .../httpd/rpc/project/ProjectAccessFactory.java    |   17 +----
 .../httpd/rpc/project/ProjectAdminServiceImpl.java |   11 +++-
 .../gerrit/httpd/rpc/project/ProjectAncestors.java |   78 ++++++++++++++++++++
 .../gerrit/httpd/rpc/project/ProjectModule.java    |    1 +
 8 files changed, 118 insertions(+), 38 deletions(-)
 create mode 100644 gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectAncestors.java

diff --git a/gerrit-common/src/main/java/com/google/gerrit/common/data/ProjectAccess.java b/gerrit-common/src/main/java/com/google/gerrit/common/data/ProjectAccess.java
index 2a80d52..365fb38 100644
--- a/gerrit-common/src/main/java/com/google/gerrit/common/data/ProjectAccess.java
+++ b/gerrit-common/src/main/java/com/google/gerrit/common/data/ProjectAccess.java
@@ -22,7 +22,6 @@ import java.util.Set;
 public class ProjectAccess {
   protected Project.NameKey projectName;
   protected String revision;
-  protected Project.NameKey inheritsFrom;
   protected List<AccessSection> local;
   protected Set<String> ownerOf;
   protected boolean isConfigVisible;
@@ -46,14 +45,6 @@ public class ProjectAccess {
     revision = name;
   }
 
-  public Project.NameKey getInheritsFrom() {
-    return inheritsFrom;
-  }
-
-  public void setInheritsFrom(Project.NameKey name) {
-    inheritsFrom = name;
-  }
-
   public List<AccessSection> getLocal() {
     return local;
   }
diff --git a/gerrit-common/src/main/java/com/google/gerrit/common/data/ProjectAdminService.java b/gerrit-common/src/main/java/com/google/gerrit/common/data/ProjectAdminService.java
index 4d10a42..c3f3d30 100644
--- a/gerrit-common/src/main/java/com/google/gerrit/common/data/ProjectAdminService.java
+++ b/gerrit-common/src/main/java/com/google/gerrit/common/data/ProjectAdminService.java
@@ -36,14 +36,17 @@ public interface ProjectAdminService extends RemoteJsonService {
   void projectDetail(Project.NameKey projectName,
       AsyncCallback<ProjectDetail> callback);
 
+  void projectAncestors(Project.NameKey projectName,
+      AsyncCallback<List<Project.NameKey>> callback);
+
+  void projectAccess(Project.NameKey projectName,
+      AsyncCallback<ProjectAccess> callback);
+
   @SignInRequired
   void createNewProject(String projectName, String parentName,
       boolean emptyCommit, boolean permissionsOnly,
       AsyncCallback<VoidResult> callback);
 
-  void projectAccess(Project.NameKey projectName,
-      AsyncCallback<ProjectAccess> callback);
-
   @SignInRequired
   void changeProjectSettings(Project update,
       AsyncCallback<ProjectDetail> callback);
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/ProjectAccessEditor.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/ProjectAccessEditor.java
index e6e11b0..0dde87e 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/ProjectAccessEditor.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/ProjectAccessEditor.java
@@ -16,6 +16,7 @@ package com.google.gerrit.client.admin;
 
 import com.google.gerrit.client.Dispatcher;
 import com.google.gerrit.client.Gerrit;
+import com.google.gerrit.client.rpc.GerritCallback;
 import com.google.gerrit.client.ui.Hyperlink;
 import com.google.gerrit.common.data.AccessSection;
 import com.google.gerrit.common.data.GitwebLink;
@@ -53,7 +54,7 @@ public class ProjectAccessEditor extends Composite implements
   DivElement inheritsFrom;
 
   @UiField
-  Hyperlink parentProject;
+  FlowPanel ancestors;
 
   @UiField
   DivElement history;
@@ -100,14 +101,26 @@ public class ProjectAccessEditor extends Composite implements
 
     this.value = value;
 
-    Project.NameKey parent = value.getInheritsFrom();
-    if (parent != null) {
+    inheritsFrom.getStyle().setDisplay(Display.NONE);
+    Util.PROJECT_SVC.projectAncestors(value.getProjectName(),
+        new GerritCallback<List<Project.NameKey>>() {
+          @Override
+          public void onSuccess(final List<Project.NameKey> result) {
+            populateAncestors(result);
+          }
+        }
+    );
+    addSection.setVisible(value != null && editing && !value.getOwnerOf().isEmpty());
+  }
+
+  private void populateAncestors(List<Project.NameKey> list) {
+    if (list.size() != 0) {
       inheritsFrom.getStyle().setDisplay(Display.BLOCK);
-      parentProject.setText(parent.get());
-      parentProject.setTargetHistoryToken( //
-          Dispatcher.toProjectAdmin(parent, ProjectScreen.ACCESS));
-    } else {
-      inheritsFrom.getStyle().setDisplay(Display.NONE);
+      ancestors.clear();
+      for (Project.NameKey prj : list) {
+        ancestors.add(new Hyperlink(prj.get(),
+            Dispatcher.toProjectAdmin(prj, ProjectScreen.ACCESS)));
+      }
     }
 
     final GitwebLink c = Gerrit.getConfig().getGitwebLink();
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/ProjectAccessEditor.ui.xml b/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/ProjectAccessEditor.ui.xml
index 4942b83..abfec55 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/ProjectAccessEditor.ui.xml
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/admin/ProjectAccessEditor.ui.xml
@@ -55,13 +55,13 @@ limitations under the License.
 <g:HTMLPanel>
   <div ui:field='inheritsFrom' class='{style.inheritsFrom}'>
     <span class='{style.parentTitle}'><ui:msg>Rights Inherit From:</ui:msg></span>
-    <q:Hyperlink ui:field='parentProject' styleName='{style.parentLink}'/>
   </div>
   <div ui:field='history' class='{style.history}'>
     <span class='{style.historyTitle}'><ui:msg>History:</ui:msg></span>
     <g:Anchor ui:field='gitweb' styleName='{style.gitwebLink}'></g:Anchor>
   </div>
 
+  <g:FlowPanel ui:field='ancestors'/>
   <g:FlowPanel ui:field='localContainer'/>
   <div class='{style.addContainer}'>
     <g:Anchor
diff --git a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectAccessFactory.java b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectAccessFactory.java
index d2a2900..ecd62fa 100644
--- a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectAccessFactory.java
+++ b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectAccessFactory.java
@@ -178,25 +178,10 @@ class ProjectAccessFactory extends Handler<ProjectAccess> {
     }
 
     final ProjectAccess detail = new ProjectAccess();
-    detail.setProjectName(projectName);
-
+    detail.setProjectName(config.getProject().getNameKey());
     if (config.getRevision() != null) {
       detail.setRevision(config.getRevision().name());
     }
-
-    if (projectName.equals(allProjectsName)) {
-      if (pc.isOwner()) {
-        ownerOf.add(AccessSection.GLOBAL_CAPABILITIES);
-      }
-      detail.setInheritsFrom(null);
-
-    } else if (config.getProject().getParent() != null) {
-      detail.setInheritsFrom(config.getProject().getParent());
-
-    } else {
-      detail.setInheritsFrom(allProjectsName);
-    }
-
     detail.setLocal(local);
     detail.setOwnerOf(ownerOf);
     detail.setConfigVisible(pc.isOwner()
diff --git a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectAdminServiceImpl.java b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectAdminServiceImpl.java
index bf7eeae..9253350 100644
--- a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectAdminServiceImpl.java
+++ b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectAdminServiceImpl.java
@@ -43,6 +43,7 @@ class ProjectAdminServiceImpl implements ProjectAdminService {
   private final CreateProjectHandler.Factory createProjectHandlerFactory;
   private final ProjectDetailFactory.Factory projectDetailFactory;
   private final SuggestParentCandidatesHandler.Factory suggestParentCandidatesHandlerFactory;
+  private final ProjectAncestors.Factory projectAncestorsFactory;
 
   @Inject
   ProjectAdminServiceImpl(final AddBranch.Factory addBranchFactory,
@@ -55,7 +56,8 @@ class ProjectAdminServiceImpl implements ProjectAdminService {
       final ProjectAccessFactory.Factory projectAccessFactory,
       final ProjectDetailFactory.Factory projectDetailFactory,
       final SuggestParentCandidatesHandler.Factory parentCandidatesFactory,
-      final CreateProjectHandler.Factory createNewProjectFactory) {
+      final CreateProjectHandler.Factory createNewProjectFactory,
+      final ProjectAncestors.Factory projectAncestorsFactory) {
     this.addBranchFactory = addBranchFactory;
     this.changeProjectAccessFactory = changeProjectAccessFactory;
     this.changeProjectSettingsFactory = changeProjectSettingsFactory;
@@ -67,6 +69,7 @@ class ProjectAdminServiceImpl implements ProjectAdminService {
     this.projectDetailFactory = projectDetailFactory;
     this.createProjectHandlerFactory = createNewProjectFactory;
     this.suggestParentCandidatesHandlerFactory = parentCandidatesFactory;
+    this.projectAncestorsFactory = projectAncestorsFactory;
   }
 
   @Override
@@ -91,6 +94,12 @@ class ProjectAdminServiceImpl implements ProjectAdminService {
   }
 
   @Override
+  public void projectAncestors(final Project.NameKey projectName,
+      final AsyncCallback<List<Project.NameKey>> callback) {
+    projectAncestorsFactory.create(projectName).to(callback);
+  }
+
+  @Override
   public void projectAccess(final Project.NameKey projectName,
       final AsyncCallback<ProjectAccess> callback) {
     projectAccessFactory.create(projectName).to(callback);
diff --git a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectAncestors.java b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectAncestors.java
new file mode 100644
index 0000000..b901885
--- /dev/null
+++ b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectAncestors.java
@@ -0,0 +1,78 @@
+// Copyright (c) 2012, Code Aurora Forum. All rights reserved.
+
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Code Aurora Forum, Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.httpd.rpc.project;
+
+
+import com.google.gerrit.httpd.rpc.Handler;
+import com.google.gerrit.reviewdb.Project;
+import com.google.gerrit.server.project.NoSuchProjectException;
+import com.google.gerrit.server.project.ProjectControl;
+import com.google.gerrit.server.project.ProjectState;
+import com.google.inject.Inject;
+import com.google.inject.assistedinject.Assisted;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+class ProjectAncestors extends Handler<List<Project.NameKey>> {
+  interface Factory {
+    ProjectAncestors create(@Assisted Project.NameKey projectName);
+  }
+
+  private final ProjectControl.Factory projectControlFactory;
+  private final Project.NameKey projectName;
+
+  @Inject
+  ProjectAncestors(final ProjectControl.Factory projectControlFactory,
+      @Assisted final Project.NameKey projectName) {
+    this.projectControlFactory = projectControlFactory;
+
+    this.projectName = projectName;
+  }
+
+  @Override
+  public ArrayList<Project.NameKey> call() throws NoSuchProjectException {
+    LinkedHashSet ancestors = new LinkedHashSet();
+    final ProjectControl pc =
+        projectControlFactory.validateFor(projectName, ProjectControl.OWNER
+            | ProjectControl.VISIBLE);
+    ProjectState projectState = pc.getProjectState().getParentState();
+    Project.NameKey prj;
+    while (projectState != null) {
+      prj = projectState.getProject().getNameKey();
+      if (!ancestors.add(prj)) {
+        break;
+      }
+      projectState = projectState.getParentState();
+    }
+    return new ArrayList<Project.NameKey>(ancestors);
+  }
+}
diff --git a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectModule.java b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectModule.java
index 2eb55b3..5087eee 100644
--- a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectModule.java
+++ b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/project/ProjectModule.java
@@ -39,6 +39,7 @@ public class ProjectModule extends RpcServletModule {
         factory(ProjectAccessFactory.Factory.class);
         factory(ProjectDetailFactory.Factory.class);
         factory(SuggestParentCandidatesHandler.Factory.class);
+        factory(ProjectAncestors.Factory.class);
       }
     });
     rpc(ProjectAdminServiceImpl.class);
-- 
1.7.7.1

